platform/engineering/request-for-change/2020-01-08-ssoe-removing-implicit-mvi-queries.md

```diff
@@ -0,0 +1,106 @@
+
+5. **OAuth Flow - Lighthouse**
+SAML authentication and MVI resolution is also performed in the context of the Lighthouse OAuth flow. MVI is queried by an API endpoint triggered by the SAML proxy, and cached MVI data is used when serving OAuth-protected API requests.<br/>
+At this point, there is no intent to integrate SSOe with the OAuth SAML interaction. Signing in to authorize an OAuth application will not establish an IAM session.<br/>
+Current proposal for the OAuth code path is to leave the MVI caching mechanism in place. This will mean that the `User` model definitely needs to remain robust to `va_profile` being populated either from an SSOe SAML payload or from an MVI lookup for the indefinite future. We should audit all of the OAuth-protected APIs to reconcile uses of various delegated methods on `User` to make sure there are no conflicts.
+
+### Development Concern - Mock Data
+We currently mock a wide range of backend service requests in our dev environments using the Betamocks library.  This insulates dev from unreliable backend systems, but also lets us customize responses for certain users in order to simulate different response scenarios.  This mocking includes the MVI requests that are done during sign-in.
+
+With the above change, we would not easily be able to mock the MVI data. Because the SAML requests/respones are mediated by the user's browser, Betamocks is not capable of intercepting those requests. So dev would be configured to a live SSOe integration environment. This means we lose the ability to customize MVI response data, and in particular we lose any existing customizations of MVI data that have been done so far. We revert to whatever is in MVI for the SSOe environment connected to dev.
+
+It's unclear how much of a problem this poses. We _could_ mitigate this by retaining the ability to do an MVI lookup on every sign-in (e.g. via a configuration setting) and enabling that capability in dev. The MVI lookup would point at the mocked MVI service. 
+
+### Attribute Priority
+There are several user traits that are currently available in both the incoming ID.me SAML assertion and in MVI (`first, middle, last, gender, birth_date, zip, ssn`). vets-api currently prefers the values from ID.me, and only falls back to the MVI values if the ID.me values are not present. 
+
+With an SSOe-based SAML payload we don't have this conflict; we'll only get one value for each of these traits. Presumably this will be the MVI value. On paper this is more "correct" as MVI is the canonical source for VA identity. But we know there are cases where this data is less accurate. VA.gov loses the ability to make this trade-off and prefer one source of another; it devolves to the IAM team to address (and ideally resolve) data mismatches and fidelity issues. 
```

Comment:
reading this part scares me a bit --- it seems like we're likely have a one-time cutover then where a bunch of the data might be wrong, since we're dropping ID.me as a source-of-truth of this data? what can we do to mitigate this ahead of time?

Reply:
I can ask if IAM does anything today when they see an incoming identity assertion from ID.me and there's any data discrepancy. 
FWIW I know vets-api flags some discrepancies but it's mostly been in the aid of debugging, we don't have any mechanism in place to follow up/correct data in MVI if there is a discrepancy.